src/riscv/lib/src/state_backend/proof_backend/proof/deserialise_stream.rs

```diff
@@ -79,27 +76,116 @@ impl StreamInput<'_> {
     }
 }
 
-/// Deserialiser for [`Deserialiser`] based on a stream of bytes.
-#[derive(Clone)]
-pub enum StreamDeserialiser<'t> {
+/// Current deserialisation context with regards to presence of the parent node
+enum Presence {
+    /// Absent node context
+    ///
+    /// This is because the parent node is either absent or blinded. In both cases, the potential
+    /// child nodes are not contained in the proof, hence we consider them absent for the purpose
+    /// of deserialisation.
     Absent,
-    Present { tags: Rc<RefCell<TagIter<'t>>> },
+
+    /// Present node context
+    ///
+    /// The parent is contained in the proof, hence we make no assumptions on the existence of
+    /// child nodes.
+    Present,
+}
+
+/// Deserialiser for [`Deserialiser`] based on a stream of bytes.
+pub struct StreamDeserialiser<'t> {
+    /// Context layers with regard to presence
+    ///
+    /// The last element describes the current node context. The first element describes the root
+    /// context.
+    ///
+    /// Contextual presence information is needed to allow moving the tag source through the
+    /// different combinators without losing track of whether the parent node is currently absent
+    /// or present.
+    context: Vec<Presence>,
+
+    /// Source of tags
+    ///
+    /// The tag source is present unconditionally because it may be passed to a parent layer. That
+    /// is because we require the ability to recover the tag source from an combinator for an
+    /// absent node.
+    tags: TagIter<'t>,
+}
+
+impl<'t> StreamDeserialiser<'t> {
+    /// Create a new deserialiser for a present context with the given tags.
+    pub fn new_present(tags: TagIter<'t>) -> Self {
+        StreamDeserialiser {
+            context: vec![Presence::Present],
+            tags,
+        }
+    }
+
+    /// Create a new deserialiser for an absent context with the given tags.
+    pub fn new_absent() -> Self {
+        StreamDeserialiser {
+            context: vec![Presence::Absent],
+            tags: TagIter::new(&[]),
+        }
+    }
+
+    /// Obtain next tag based on the current state of the deserialiser.
+    ///
+    /// Return [`None`] if the parent node is absent / blinded - making it nonsensical
+    /// to obtain any tags.  
+    fn next_tag(&mut self) -> Option<Result<Tag, TagError>> {
```

Comment:
Somewhat unrelated to this PR since the signature of this function is copied over, but this could be easily turned into a `Result<Tag, TagError>` with an additional `TagError` variant.

Reply:
I think there is no need for an extra variant. The absence can be determined upstream, which is where it is handled anyway. I lightly refactored how `next_tag()` is called in [this diff](https://github.com/tezos/riscv-pvm/compare/d353193d575b84f22320b7369d024eb1b9ea9505..156f1d8f539e68ee1252eb4fa4ff5659ac5db4d2).

Reply:
The doc comment is out of date now.

Reply:
I corrected it [here](https://github.com/tezos/riscv-pvm/compare/156f1d8f539e68ee1252eb4fa4ff5659ac5db4d2..a2afc71bd35c67b8004052f5942694efbf6d0cd1).